### PR TITLE
feat($core): Improve VuePress build time

### PR DESCRIPTION
**Pull request by kefranabg**
**Summary**

These optimisations make sense on projects that contains a huge amount of pages to render. 

Related to #1560 

**What kind of change does this PR introduce?** (check at least one)

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update
- [ ] Refactor
- [ ] Docs
- [ ] Build-related changes
- [x] Other, please describe:
Performance enhancement

**Does this PR introduce a breaking change?** (check one)

- [ ] Yes
- [x] No

**Steps:**

[Improvement #1](https://github.com/vuejs/vuepress/commit/9dce83575a8e798ee29fa8ddedacfdfbbb3e7ac6) : Render static pages in parallel instead of rendering one page after the other.
[Improvement #2](https://github.com/vuejs/vuepress/pull/2163/commits/cbe1af0b674d209beedb91c7a4692effc3118d5a) : Run extendPageData enhancers in parallel

### CODE REVIEW

**Comment by meteorlxy** [comment]
Tip: You can use "Create draft PR" for WIP PRs

**Comment by kefranabg** [comment]
@meteorlxy Yes I forgot about this one 😅 👍

**Comment by TribeWeb** [comment]
Hi @kefranabg. I have a project where the build was gradually slowing down until it fails (blows the stack) and I just tried your new code. It now renders all the pages in less than a second!

**Comment by itsxallwater** [comment]
@TribeWeb I'd be curious to know how many docs in your project and if you ran this code on top of 1.2 or 1.3. I've got a project with 2k+ docs and dropping this code onto 1.3, I'm still seeing lengthy build times (15 minutes and counting at the moment). I also was getting "JavaScript heap out of memory" errors mid-build until I changed the build script to set node's `max_old_space_size` to 8192.

Keen to hear a bit more as I was working on a separate approach to pull in node's worker_threads which I had working under 1.2 and had dropped my build time from 1 hour to 7.5 minutes, though the move to 1.3 has created some issues.

**Comment by TribeWeb** [comment]
> I'd be curious to know how many docs in your project and if you ran this code on top of 1.2 or 1.3

Hi @itsxallwater. It’s a much smaller project about 100 pages. I’m using it with Vuetify and moving from v1.5 to 2.2. Build time was seconds previous but now hitting out of memory errors until I tried the fix above. VuePress is 1.3. I keep hitting this problem sometimes for weird reasons. I isolated the build time problem on the previous version of Vuetify to one button and fixed by wrapping it in <ClientOnly>.

**Comment by bencodezen** [comment]
Great work on this @kefranabg!

**Comment by meteorlxy** [comment]
About improvement #1, https://github.com/vuejs/vuepress/pull/315#issuecomment-386339054 discuessed long time ago (in v0.x 😅). Let's check if it would break the cli output in current version.

**Comment by meteorlxy** [comment]
Oh, so you removed the cli output in the PR.

Have you made some benchmark for these improvements? @kefranabg

**Comment by kefranabg** [comment]
@meteorlxy I'll update the description later to add the benchmark 😄On small projects like the vuepress doc, these changes are not that effective. But on projects that have like 600pages to render, it has an interesting impact.

There is still some improvements to do to improve the build time, probably on webpack config.

**Comment by kefranabg** [comment]
@meteorlxy

With more than 600 pages to render

**Without optimization:**
build time = 360 seconds
**With optimizations:**
build time = 120 seconds

File: packages/@vuepress/core/lib/node/Page.js [hunk 1]
```diff
@@ -286,13 +286,20 @@ module.exports = class Page {
...
+        console.log(e)
+        return new Error(`[${pluginName}] execute extendPageData failed.`)
+      })
+    })
+
+    const results = await Promise.all(enhancerPromises)
+    const enhancerError = results.find(result => result instanceof Error)
+    if (enhancerError) throw enhancerError
+
+    return results
```
Comment (meteorlxy) [inline packages/@vuepress/core/lib/node/Page.js]:
```suggestion
    return Promise.all(enhancers.map(async ({ value: enhancer, name: pluginName }) => {
      try {
        await enhancer(this)
      } catch (error) {
        console.log(error)
        throw new Error(`[${pluginName}] execute extendPageData failed.`)
      }
    }))
```

What about this?

File: packages/@vuepress/core/lib/node/build/index.js [hunk 2]
```diff
@@ -89,11 +89,13 @@ module.exports = class Build extends EventEmitter {
...
     logger.wait('Rendering static HTML...')
 
-    const pagePaths = []
+    const pagePathsPromises = []
     for (const page of this.context.pages) {
-      pagePaths.push(await this.renderPage(page))
+      pagePathsPromises.push(this.renderPage(page))
     }
 
+    const pagePaths = await Promise.all(pagePathsPromises)
```
Comment (meteorlxy) [inline packages/@vuepress/core/lib/node/build/index.js]: From 91 to 97:

```suggestion
    const pagePaths = await Promise.all(
      this.context.pages.map(page => this.renderPage(page))
    )
```

File: packages/@vuepress/core/lib/node/Page.js (hunk 1 above)
Comment (kefranabg) [reply packages/@vuepress/core/lib/node/Page.js]: 😅 way better

**Comment by Mister-Hope** [comment]
Have you checked #2221? I am afraid this change will make this issue even worse.@kefranabg

**Comment by itsxallwater** [comment]
FWIW these updates took our build time down to ~10 minutes, and we're talking ~1500 docs with these plugins:

image

@Mister-Hope might be noteworthy but we're overloading the build command in the package.json to run as `node --max_old_space_size=8192 ./node_modules/vuepress/cli.js build` allowing node to use more memory.

**Comment by Mister-Hope** [comment]
> FWIW these updates took our build time down to ~10 minutes, and we're talking ~1500 docs with these plugins:
> 
> image
> 
> @Mister-Hope might be noteworthy but we're overloading the build command in the package.json to run as `node --max_old_space_size=8192 ./node_modules/vuepress/cli.js build` allowing node to use more memory.
I knew node can use more memory. But I am talking about memory leak. Before changing, I don't think it's normal that the memory keep rising during build.
@itsxallwater

**Comment by itsxallwater** [comment]
I don't disagree, just wanted to add context from our build experience.

**Comment by Mister-Hope** [comment]
> I don't disagree, just wanted to add context from our build experience.

Let me explain my idea more carefully, I agree that the build process of pages should be parrell intead of await each other. But there must be some GC issues while building. Your FR is nice, since the offcial accepted it, I want to remind the offical that this may make the GC issue even worse.

At least I am using even more memory in 1.4.0.

**Comment by itsxallwater** [comment]
I'm not arguing :)

More context for the thread at large--we've moved our build to a GitHub action and it's reliably coming in at 20 minutes.

**Pull request merged**